### PR TITLE
Setup workspace and feature switching

### PR DESCRIPTION
**Pull request by ascobie**
Fix: #1449 

- [x] Decide on JSON template structure
- [x] Create master or admin template with all features enabled
- [x] Read workspace templates
- [x] Create workspace selector DDL
- [x] Feature directive for hiding/showing components

Note: ability to override workspace settings can be added in another PR and feature request as this one is already big enough.

### CODE REVIEW

**Comment by ascobie** [comment]
So just pass it a workspace template file contents without the user overrides? Then it would parse it and set the Id and name?

Get Outlook for Android<https://aka.ms/ghei36>

File: app/components/workspace/json-templates/admin-workspace.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+    "metadata": {
```
Comment (timotheeguerin) [inline app/components/workspace/json-templates/admin-workspace.json]: why not move all of those up one level if we have workspace bellow. Also `features` might be better?

File: app/components/workspace/workspace-selector/workspace-dropdown.component.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        private workspaceService: WorkspaceService) {
+
+        this.workspaces = this.workspaceService.workspaces;
+        this._subscriptions.push(this.workspaces.subscribe((workspaces) => {
+            console.log("workspaces.subscribe: ", workspaces);
+            this.loaded = true;
+            this.changeDetector.markForCheck();
+        }));
+
+        this._subscriptions.push(workspaceService.currentWorkspace.subscribe((workspace: Workspace) => {
```
Comment (timotheeguerin) [inline app/components/workspace/workspace-selector/workspace-dropdown.component.ts]: you probably don't need this one here. This was for accounts

File: app/services/workspace.service.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import { List } from "immutable";
+import { BehaviorSubject, Observable } from "rxjs";
+
+// tslint:disable:no-var-requires max-line-length
+const stripJsonComments = require("strip-json-comments");
+
+import { Workspace } from "app/models";
+
+// TODO: can we just read all JSON files in a directory?
+const adminWorkspace = JSON.parse(stripJsonComments(require("app/components/workspace/json-templates/admin-workspace.json")));
```
Comment (timotheeguerin) [inline app/services/workspace.service.ts]: Let's start like that, we'll look into custom user workspace later. You can also remove the stripJsonComments here. This file shouldn't have some

File: app/components/workspace/json-templates/admin-workspace.json (hunk 1 above)
Comment (ascobie) [reply app/components/workspace/json-templates/admin-workspace.json]: if you are happy with that then we could do. didn't want metadata mixing with the actual control values.

File: app/components/workspace/workspace-selector/workspace-dropdown.component.ts (hunk 2 above)
Comment (ascobie) [reply app/components/workspace/workspace-selector/workspace-dropdown.component.ts]: yeah this was just to test that the current workspace was being set and subscribable. will remove.

File: .vscode/settings.json [hunk 4]
```diff
@@ -26,5 +26,9 @@
   "tslint.alwaysShowRuleFailuresAsWarnings": true,
   "css.validate": false,
   "scss.validate": false,
-  "python.formatting.provider": "yapf"
+  "python.formatting.provider": "yapf",
+  "[json]": {
```
Comment (timotheeguerin) [inline .vscode/settings.json]: can you revert that? We use .editorConfig for those settings and json is set to 2 space on purpose as npm will use 2 space for package.json

**Comment by ascobie** [comment]
Ok, I can revert it. Though viewing the json templates in VS Code is very annoying at 2 spaces. Is there any other way I can change it? I can't read json with 2 spaces.



Get Outlook for Android<https://aka.ms/ghei36>

________________________________

File: app/app.component.ts [hunk 5]
```diff
@@ -51,7 +51,11 @@ export class AppComponent implements OnInit {
     ) {
         this.autoscaleFormulaService.init();
         this.settingsService.init();
+<<<<<<< HEAD
```
Comment (timotheeguerin) [inline app/app.component.ts]: merge issue

File: app/components/common/feature-switching/index.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export * from "../../../../src/@batch-flask/ui/workspace/feature-visible.directive";
```
Comment (timotheeguerin) [inline app/components/common/feature-switching/index.ts]: remove this file

File: app/components/job-schedule/details/job-schedule-configuration.component.ts [hunk 7]
```diff
@@ -22,8 +24,26 @@ export class JobScheduleConfigurationComponent {
...
 
     private _jobSchedule: JobSchedule;
+    private _sub: Subscription;
+
+    constructor(
+        private changeDetector: ChangeDetectorRef,
+        private workspaceService: WorkspaceService) {
+
+        this._sub = this.workspaceService.currentWorkspace.subscribe((ws) => {
+            if (ws) {
```
Comment (timotheeguerin) [inline app/components/job-schedule/details/job-schedule-configuration.component.ts]: i feel like we should always assume this is never null. We'll make sure there is always a value in the service

File: app/components/layout/main-navigation/main-navigation.html [hunk 8]
```diff
@@ -2,27 +2,41 @@
...
-        <i class="fa fa-certificate fa-lg"></i><div class="label">Certificate</div>
-    </bl-clickable>
-    <bl-clickable *blHiddenIfNoAccount routerLink="/data" routerLinkActive="active" title="File groups">
-        <i class="fa fa-cloud-upload fa-lg"></i><div class="label">Data</div>
-    </bl-clickable>
-    <bl-clickable *blHiddenIfNoAccount href="#" routerLink="market" routerLinkActive="active" title="Gallery">
-        <i class="fa fa-cubes" aria-hidden="true"></i><div class="label">Gallery</div>
-    </bl-clickable>
+    <div *blHiddenIfNoAccount>
+        <bl-clickable *blFeatureOn="'job.view'" routerLink="/jobs" routerLinkActive="active">
```
Comment (timotheeguerin) [inline app/components/layout/main-navigation/main-navigation.html]: shouldn't all of those just be `job`?

File: app/components/workspace/json-templates/end-user-workspace.json [hunk 9]
```diff
@@ -0,0 +1,50 @@
+{
+    "id": "end-user.workspace",
+    "displayName": "Basic end-user workspace",
+    "description": "Default end-user workspace that hides a number of admin related features.",
```
Comment (timotheeguerin) [inline app/components/workspace/json-templates/end-user-workspace.json]: I would not call admin(Maybe advanced?)

File: app/components/workspace/json-templates/end-user-workspace.json [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                    "json": false
+                }
+            },
+            "action": {
+                "clone": false,
+                "exportAsJson": false,
+                "pin": false
+            }
+        },
+        "schedule": {
```
Comment (timotheeguerin) [inline app/components/workspace/json-templates/end-user-workspace.json]: feel like we could move all the view one step above. It doesn't really make sense to only disable view

```
{
 "job": {
   "statistics": true,
   "tags": true,
 }
}
```

File: src/@batch-flask/ui/entity-commands/entity-command.ts [hunk 11]
```diff
@@ -17,6 +18,10 @@ export enum EntityCommandNotify {
 }
 
 export interface EntityCommandAttributes<TEntity extends ActionableEntity, TOptions = void> {
+    /**
+     * Name is to be used by feature to show buttons. Or to define keyboard shortcuts
+     */
+    name?: string;
```
Comment (timotheeguerin) [inline src/@batch-flask/ui/entity-commands/entity-command.ts]: remove the ? for this one. I added it so it wouldn't break everything but now all commands should have a name(Add to those missing)

File: src/@batch-flask/ui/workspace/feature-visible.directive.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    constructor(
+        private templateRef: TemplateRef<any>,
+        private viewContainer: ViewContainerRef,
+        private workspaceService: WorkspaceService,
+    ) {
+    }
+
+    public ngOnInit() {
+        this._subscription = this.workspaceService.currentWorkspace.subscribe((workspace) => {
+            if (this.feature && workspace) {
```
Comment (timotheeguerin) [inline src/@batch-flask/ui/workspace/feature-visible.directive.ts]: you need to listen to the `ngOnChanges` here too otherwise if feature changes then it won't update

File: src/@batch-flask/ui/workspace/workspace.model.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            if (definition === true) {
+                return true;
+            } else if (definition === undefined) {
+                return true;
+            } else if (definition === false) {
+                return false;
+            }
+        }
+
+        // TODO: debug - remove
```
Comment (timotheeguerin) [inline src/@batch-flask/ui/workspace/workspace.model.ts]: remove

File: src/@batch-flask/ui/workspace/workspace.model.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        if (definition) {
+            // console.log(`traversed feature '${requestedFeature}' to end ... definition tree ->`, definition);
+        }
+
+        // if there are any remaining template definition children after we
+        // have got to the end of the requested feature, we can assume this means
+        // that it should be enabled.
+        return definition ? true : false;
```
Comment (timotheeguerin) [inline src/@batch-flask/ui/workspace/workspace.model.ts]: think this should be definition == false ? false : true

File: src/@batch-flask/ui/workspace/workspace.service.spec.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+describe("WorkspaceService", () => {
+    // TODO
```
Comment (timotheeguerin) [inline src/@batch-flask/ui/workspace/workspace.service.spec.ts]: Todo tests

File: app/components/workspace/json-templates/end-user-workspace.json (hunk 10 above)
Comment (ascobie) [reply app/components/workspace/json-templates/end-user-workspace.json]: Action can stay where it is I reckon.

File: src/@batch-flask/ui/workspace/feature-visible.directive.ts (hunk 12 above)
Comment (ascobie) [reply src/@batch-flask/ui/workspace/feature-visible.directive.ts]: Why would the feature change?

Comment (timotheeguerin) [reply src/@batch-flask/ui/workspace/feature-visible.directive.ts]: why not, might be some computed features

Comment (ascobie) [reply src/@batch-flask/ui/workspace/feature-visible.directive.ts]: Yeah I guess that's true. Future proofing is not a bad idea.

File: app/components/workspace/json-templates/end-user-workspace.json (hunk 10 above)
Comment (nehal-j) [reply app/components/workspace/json-templates/end-user-workspace.json]: After removing view, I need to rename the tags because 'tags' has been used for the summaryTags as well. 
<bl-tags *blFeatureOn="'pool.paper'" summaryTags [tags]="pool.tags" [editable]="true" [save]="updateTags" noTagsMessage="No tags."></bl-tags>

File: src/@batch-flask/ui/workspace/workspace.service.spec.ts (hunk 15 above)
Comment (ascobie) [reply src/@batch-flask/ui/workspace/workspace.service.spec.ts]: Nehal working on those now.

File: app/app.component.ts [hunk 16]
```diff
@@ -4,7 +4,8 @@ import { DomSanitizer } from "@angular/platform-browser";
 import { Observable } from "rxjs";
 
 import { ActivatedRoute } from "@angular/router";
-import { IpcService } from "@batch-flask/ui";
+import { IpcService, Workspace, WorkspaceService } from "@batch-flask/ui";
+// import { MonacoLoader } from "@batch-flask/ui/editor";
```
Comment (timotheeguerin) [inline app/app.component.ts]: remove

File: app/components/data/shared/job-id/job-id.component.ts [hunk 17]
```diff
@@ -70,7 +70,12 @@ export class JobIdComponent implements AsyncValidator, ControlValueAccessor, OnD
 
     @autobind()
     private _validateJobUnique(control: FormControl) {
-        return Observable.of(null).debounceTime(250)
+        // on init value is an empty array.
+        if (!control.value || (Array.isArray(control.value) && control.value.length === 0)) {
+            return Observable.of(null);
```
Comment (timotheeguerin) [inline app/components/data/shared/job-id/job-id.component.ts]: change to `of(null)` in rxjs 6

File: app/components/job/action/job-commands.ts [hunk 18]
```diff
@@ -189,4 +197,12 @@ export class JobCommands extends EntityCommands<Job> {
             return Observable.fromPromise(this.fs.saveFile(localPath, content));
```
Comment (timotheeguerin) [inline app/components/job/action/job-commands.ts]: `Observable.fromPromise` change to `from` in genera change all Observable. to the rxjs 6 version

File: app/app.component.ts [hunk 19]
```diff
@@ -5,7 +5,6 @@ import { Observable } from "rxjs";
 
 import { ActivatedRoute } from "@angular/router";
 import { IpcService, Workspace, WorkspaceService } from "@batch-flask/ui";
-// import { MonacoLoader } from "@batch-flask/ui/editor";
 import { PermissionService } from "@batch-flask/ui/permission";
```
Comment (ascobie) [inline app/app.component.ts]: still don't know how this made its way in. i think you should revert this change.

File: app/components/data/shared/job-id/job-id.component.spec.ts [hunk 20]
```diff
@@ -1,81 +1,86 @@
-import { Component, DebugElement, NO_ERRORS_SCHEMA } from "@angular/core";
```
Comment (ascobie) [inline app/components/data/shared/job-id/job-id.component.spec.ts]: commented out this entire fixture due to it breaking the build with the recent addition of the "timer(delay)". need Tim's assistance to fix it.

**Comment by ascobie** [comment]
Just merged master into this branch. There now seems to be an issue with i18n and it not computing when the component is re-added: 

image

Any ideas how to fix this?

File: app/services/file/file-navigator/file-tree.model.spec.ts [hunk 21]
```diff
@@ -15,6 +15,7 @@ function reprTree(tree: FileTreeStructure) {
 
 function reprNode(node: FileTreeNode, indent = 0) {
     const rows = [];
+    // tslint:disable-next-line:no-unused-variable
     for (const [_, child] of node.children) {
```
Comment (ascobie) [inline app/services/file/file-navigator/file-tree.model.spec.ts]: the _ was failing the build, might have been a package update as this hasn't change for a long time.

Comment (timotheeguerin) [reply app/services/file/file-navigator/file-tree.model.spec.ts]: remove _ just `[,child]`
